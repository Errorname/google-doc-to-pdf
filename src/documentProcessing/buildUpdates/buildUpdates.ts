import { Placeholder, ContentPlaceholder } from '../types'
import { Unit, Request } from '../gdocTypes'

const buildUpdates = (placeholders: Placeholder[]): Request[] => {
  const updates: Request[] = []

  for (let placeholder of placeholders) {
    // ContentPlaceholder
    if (placeholder.type == 'content') {
      const contentPlaceholder = placeholder as ContentPlaceholder
      // Image
      if (contentPlaceholder.output.url) {
        const output = contentPlaceholder.output
        // Insert image
        updates.push({
          insertInlineImage: {
            uri: output.url,
            objectSize: {
              width: output.width ? { magnitude: output.width, unit: Unit.PT } : undefined,
              height: output.height ? { magnitude: output.height, unit: Unit.PT } : undefined
            },
            location: {
              segmentId: '', // Empty means body
              index: placeholder.position.start
            }
          }
        })
        // Remove mustaches
        updates.push({
          replaceAllText: {
            replaceText: '',
            containsText: {
              text: placeholder.raw,
              matchCase: true
            }
          }
        })
      }

      // Text
      if (
        contentPlaceholder.output === `${contentPlaceholder.output}` ||
        !isNaN(contentPlaceholder.output)
      ) {
        updates.push({
          replaceAllText: {
            replaceText: `${contentPlaceholder.output}`,
            containsText: {
              text: placeholder.raw,
              matchCase: true
            }
          }
        })
      }
    }

    // CommentPlaceholder
    else if (placeholder.type == 'comment') {
      updates.push({
        replaceAllText: {
          replaceText: '',
          containsText: {
            text: placeholder.raw,
            matchCase: true
          }
        }
      })
    }
  }

  // Sort updates
  /* For Google Doc, updates are applied one after the other. This means that the origin
  position of placeholders may change if a previous update was applied, (eg: with a long text)!
  We need a better dynamic position system to handle this, but in the meantime,
  we simply sort the imageUpdates first to mitigate the problem. */
  updates.sort((a, b) => {
    if (a.insertInlineImage && !b.insertInlineImage) return -1
    if (!a.insertInlineImage && b.insertInlineImage) return 1
    return 0
  })

  return updates
}

export default buildUpdates
