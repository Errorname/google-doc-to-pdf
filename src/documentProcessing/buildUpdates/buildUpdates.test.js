import buildUpdates from './buildUpdates'
describe('buildUpdates', () => {
  test('no placeholders', () => {
    const updates = buildUpdates([])

    expect(updates).toEqual([])
  })

  test('contentPlaceholder - text', () => {
    const contentPlaceholder = {
      raw: '{{ user.name.first }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'user.name.first',
        value: 'Thibaud'
      },
      pipes: [],
      output: 'Thibaud'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - text (number)', () => {
    const contentPlaceholder = {
      raw: '{{ accounts[0].money }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'accounts[0].money',
        value: 1500
      },
      pipes: [],
      output: '1500'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat'
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width)', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image(150) }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat',
        width: 150
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - image (width, height)', () => {
    const contentPlaceholder = {
      raw: '{{ catUrl | image(150, 80) }}',
      position: {
        start: 42,
        end: 47
      },
      type: 'content',
      input: {
        raw: 'catUrl',
        value: 'http://example.cat'
      },
      pipes: [],
      output: {
        url: 'http://example.cat',
        width: 150,
        height: 180
      }
    }

    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('contentPlaceholder - unknown', () => {
    const contentPlaceholder = {
      raw: '{{ strange.type }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'content',
      input: {
        raw: 'strange.type',
        value: { test: 123 }
      },
      pipes: [],
      output: { test: 123 }
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('commentPlaceholder', () => {
    const contentPlaceholder = {
      raw: '{{ # This is a comment }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'comment'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })

  test('unknownPlaceholder', () => {
    const contentPlaceholder = {
      raw: '{{ % Unknown placeholder }}',
      position: {
        start: 0,
        end: 0
      },
      type: 'not_real_type'
    }
    const updates = buildUpdates([contentPlaceholder])

    expect(updates).toMatchSnapshot()
  })
})
