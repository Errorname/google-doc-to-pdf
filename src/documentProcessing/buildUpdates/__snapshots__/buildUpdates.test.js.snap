// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildUpdates commentPlaceholder 1`] = `
Array [
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ # This is a comment }}",
      },
      "replaceText": "",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - image (width) 1`] = `
Array [
  Object {
    "insertInlineImage": Object {
      "location": Object {
        "index": 42,
        "segmentId": "",
      },
      "objectSize": Object {
        "height": undefined,
        "width": Object {
          "magnitude": 150,
          "unit": "PT",
        },
      },
      "uri": "http://example.cat",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ catUrl | image(150) }}",
      },
      "replaceText": "",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - image (width, height) 1`] = `
Array [
  Object {
    "insertInlineImage": Object {
      "location": Object {
        "index": 42,
        "segmentId": "",
      },
      "objectSize": Object {
        "height": Object {
          "magnitude": 180,
          "unit": "PT",
        },
        "width": Object {
          "magnitude": 150,
          "unit": "PT",
        },
      },
      "uri": "http://example.cat",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ catUrl | image(150, 80) }}",
      },
      "replaceText": "",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - image 1`] = `
Array [
  Object {
    "insertInlineImage": Object {
      "location": Object {
        "index": 42,
        "segmentId": "",
      },
      "objectSize": Object {
        "height": undefined,
        "width": undefined,
      },
      "uri": "http://example.cat",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ catUrl | image }}",
      },
      "replaceText": "",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - text (number) 1`] = `
Array [
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ accounts[0].money }}",
      },
      "replaceText": "1500",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - text 1`] = `
Array [
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ user.name.first }}",
      },
      "replaceText": "Thibaud",
    },
  },
]
`;

exports[`buildUpdates contentPlaceholder - unknown 1`] = `Array []`;

exports[`buildUpdates unknownPlaceholder 1`] = `Array []`;
